File: AdvancedTypes/src/main/java/com/bennyhuo/kotlin/advancedtypes/constructors/java/Person.java
package com.bennyhuo.kotlin.advancedtypes.constructors.java;

public class Person {
    private int age;
    private String name;

    public Person(int age, String name) {
        this.age = age;
        this.name = name;
    }

    {
        System.out.println("1");
    }

    {
        System.out.println("2");
    }

}

File: AdvancedTypes/src/main/java/com/bennyhuo/kotlin/advancedtypes/constructors/kotlin/Person.kt
package com.bennyhuo.kotlin.advancedtypes.constructors.kotlin


abstract class Animal
class Person(var age: Int, var name: String = "unknown") {
    override fun equals(other: Any?) = (other as? Person)?.name?.equals(name) ?: false
    override fun hashCode() = name.hashCode()
}

val persons = HashMap<String, Person>()
fun Person(name: String): Person {
    return persons[name]
        ?: Person(1, name).also { persons[name] = it }
}